src/core/vector/qgsvectorlayerundocommand.cpp

```diff
@@ -228,8 +228,17 @@ void QgsVectorLayerUndoCommandChangeAttribute::undo()
   {
     // added feature
     const QgsFeatureMap::iterator it = mBuffer->mAddedFeatures.find( mFid );
-    Q_ASSERT( it != mBuffer->mAddedFeatures.end() );
-    it.value().setAttribute( mFieldIndex, mOldValue );
+    if ( it == mBuffer->mAddedFeatures.end() )
+    {
+      // the feature must have been removed, nothing to undo here
+      // See https://github.com/qgis/QGIS/issues/23243
+      Q_ASSERT( it != mBuffer->mAddedFeatures.end() );
+    }
+    else
+    {
+      //Q_ASSERT( it != mBuffer->mAddedFeatures.end() );
+      it.value().setAttribute( mFieldIndex, mOldValue );
+    }
```

Comment:
I would remove the assertions to remove confusion.
Furthermore the Q_ASSERT( it != mBuffer->mAddedFeatures.end() ); would always evaluate to Q_ASSERT(False), and thus trigger.

```suggestion
    if ( it == mBuffer->mAddedFeatures.end() )
    {
      // the feature must have been removed, nothing to undo here
      // See https://github.com/qgis/QGIS/issues/23243
    }
    else
    {
      it.value().setAttribute( mFieldIndex, mOldValue );
    }
```

Reply:
Yes @rouault thanks for looking. I left the crash in to prove a testcase would be hitting the bug being fixed. This is a work in progress (thus the WIP in description) specifically because I'm trying to craft a test triggering the failure. I suspect it's a misuse of QUndoStack to have events that are not really completed in the stack .